unity/Holo/Assets/Scripts/STLConverter.cs

```diff
@@ -0,0 +1,178 @@
+
+        STLFileImporter stlFileImporter = new STLFileImporter();
+
+        bool first_mesh = true;
+        Debug.Log("Doing mesh #1!");
+        stlFileImporter.Load_STL_file(file_paths[0], first_mesh);
+        mesh.vertices = stlFileImporter.BaseVertices;
+        mesh.triangles = stlFileImporter.Indices;
+        mesh.AddBlendShapeFrame(Path.GetFileName(file_paths[0]), 100f, stlFileImporter.AllVertices, null, null);
+        first_mesh = false;
+
+        int count = 2;
+        for (int i = 1; i < file_paths.Length; i++)
+        {
+            Debug.Log("Doing mesh #" + count.ToString() + "!");
+            stlFileImporter.Load_STL_file(file_paths[1], first_mesh);
+            mesh.AddBlendShapeFrame(Path.GetFileName(file_paths[i]), 100f, stlFileImporter.AllVertices, null, null);
+            first_mesh = false;
+            count++;
+        }
+
+        skinnedMesh.sharedMesh = mesh;
+    }
+}
+
+
+//Loads a single STL file and turns it into a list of vertices (x,y,z) & if first_mesh: a list of indexes
+public class STLFileImporter
+{
+    public Vector3[] BaseVertices { get; private set; }
+
+    private List<Vector3> allVertices = new List<Vector3>();
+    public Vector3[] AllVertices { get => allVertices.ToArray(); }
+
+    private List<int> indices = new List<int>();
+    public int[] Indices { get => indices.ToArray(); }
+
+    private uint facetCount = 1;
+
+    public void Load_STL_file(string file_path, bool first_mesh)
+    {
+        allVertices = new List<Vector3>();
+        using (FileStream filestream = new FileStream(file_path, FileMode.Open, FileAccess.Read))
+        {
+            using (BinaryReader binary_reader = new BinaryReader(filestream, new ASCIIEncoding()))
+            {
+                // read header
+                byte[] header = binary_reader.ReadBytes(80);
+                facetCount = binary_reader.ReadUInt32();
+
+                if (first_mesh)
+                {
+                    BaseVertices = new Vector3[facetCount*3];
+                }
+
+                //TODO check topology
+
+                for (uint i = 0; i < facetCount; i++)
+                    Adapt_Facet(binary_reader, first_mesh);
+            }
+        }
+        BaseVertices = new Vector3[allVertices.Count];
+    }
+
+    private void Adapt_Facet(BinaryReader binary_reader, bool first_mesh)
+    {
+        binary_reader.GetVector3(); // A normal we don't use
+
+        for (int i = 0; i < 3; i++)
+        {
+            Vector3 vertix = binary_reader.GetVector3();
+            AddVertix(vertix, first_mesh);
+        }
+        binary_reader.ReadUInt16(); // non-sense attribute byte
+    }
+
+    private void AddVertix(Vector3 vertix, bool first_mesh)
+    {
+        int index = 0;
+        foreach (Vector3 listed_vertix in allVertices)
+        {
+            if (listed_vertix.Equals(vertix))
+            {
+                if(first_mesh)
+                    this.indices.Add(index);
+                return;
```

Comment:
I think we don't want to make here a condition `if(first_mesh)`.  That is, I think you want to just do here 

```
           if (listed_vertix.Equals(vertix))
            {
                this.indices.Add(index);
                return;
            }
```
Note that below, when the vertex is not found on the `allVertices` list, then it is added to the `this.indices` list unconditionally, regardless if it's first_mesh or not first_mesh.

It seems that AddVertix should not even need a parameter `first_mesh`. It updates `indices` and `allVertices`, always, regardless of the `first_mesh` state.

Reply:
From the TestBuildGameObject file you've sent I got the impression that we update vertices on the first mesh and then as we progress we just map vertices from the subsequent meshes to the indices of the first mesh, that's why I made it that way. (I noted that you update indices to the mesh only once). I should have been thorough and make an if statement at the end too (so indices are)
If that's the case, should I update mesh.triangles after every mesh, or just once at the beginning?

Reply:
Indeed Unity `mesh.triangles` should be set only once (all blend shapes share them).

But I thought that you may recreate `indices` in each `Load_STL_file` run, this way you can validate later that all blend shapes indeed have the same values for them. 

IOW, I would suggest this approach:
- after the 1st `Load_STL_file`, you only use Indices to set `mesh.triangles`
- after the 2nd and latter calls to `Load_STL_file`, you only use Indices to validate -- making sure that the new Indices list is equal to the initial Indices list.